protocol/src/main/java/org/fido/iot/protocol/CryptoService.java

```diff
@@ -579,6 +581,113 @@ public void verify(Composite chain) {
     }
   }
 
+  /**
+   * Verifies certificate algorithm.
+   *
+   * @param algorithmName Public key algorithm
+   */
+  public void verifyAlgorithm(String algorithmName) throws InvalidOwnershipVoucherException {
+    if (!(algorithmName.equals(Const.EC_ALG_NAME))) {
+      throw new InvalidOwnershipVoucherException(
+          "Wrong public key algorithm inside the device certificate - supported: ECDSA, received: "
+              + algorithmName);
+    }
+  }
+
+  /**
+   * Verifies certificate public key curve.
+   *
+   * @param ecdsaCurveName ECDSA curve name
+   */
+  private void verifyCurveName(String ecdsaCurveName) throws InvalidOwnershipVoucherException {
+    if (!(ecdsaCurveName.contains(Const.SECP256R1_CURVE_NAME)
+        || ecdsaCurveName.contains(Const.SECP384R1_CURVE_NAME))) {
+      throw new InvalidOwnershipVoucherException("Mismatch of ECDSA curve type.");
+    }
+  }
+
+  /**
+   * Verifies certificate public key size.
+   *
+   * @param pubKeySize Public key size
+   */
+  private void verifyKeySize(int pubKeySize) throws InvalidOwnershipVoucherException {
+    if (pubKeySize != Const.BIT_LEN_256 && pubKeySize != Const.BIT_LEN_384) {
+      throw new InvalidOwnershipVoucherException(
+          "Wrong public key size. Received public key size: {}." + pubKeySize);
+    }
+  }
+
+  /**
+   * Verifies leaf certificate public key in ownership voucher.
+   *
+   * @param cert leaf certificate from ownership voucher
+   */
+  private void verifyLeafPubKeyData(X509Certificate cert) throws InvalidOwnershipVoucherException {
+    String publicKeyAlgorithm = cert.getPublicKey().getAlgorithm();
+    verifyAlgorithm(publicKeyAlgorithm);
+
+    String ecdsaCurveName = ((ECPublicKey) cert.getPublicKey()).getParams().toString();
+    verifyCurveName(ecdsaCurveName);
+
+    int pubKeySize =
+        ((ECPublicKey) cert.getPublicKey()).getParams().getCurve().getField().getFieldSize();
+    verifyKeySize(pubKeySize);
+  }
+
+  /**
+   * Verifies leaf certificate in ownership voucher.
+   *
+   * @param cert leaf certificate from ownership voucher
+   */
+  private void verifyLeafCertPrivileges(X509Certificate cert)
+      throws InvalidOwnershipVoucherException {
+    if (cert.getKeyUsage() != null) {
+      if (!(cert.getKeyUsage()[0])) {
+        throw new InvalidOwnershipVoucherException(
+            "Digital signature is not allowed for the device certificate");
+      }
+    }
+  }
+
+  /**
+   * Verifies certificate chain.
+   *
+   * @param certChain ownership voucher device certificate chain
+   */
+  public void verifyCertChain(Composite certChain) {
```

Comment:
Functionally, it verifies the leaf cert, and not the entire cert chain, unless the method 'verify()' is called within this method. The method could be renamed in the future.